### PR TITLE
Query: Make CreateReadValueExpression an extension method

### PR DESCRIPTION
**Pull request by smitpatel**
This method is to read a particular index from valueBuffer for a property.
It was in EntityMaterializerSource as an instance method in order for provider to influence how values are read from valueBuffer.
In new query pipeline, providers are supposed to convert valueBuffer read to their database object read. Further, even if they want to create a ValueBuffer for read, the valueBuffer should have values after all custom processing.
Hence there is no particular need to make this overridable behavior. Making it extension method simplify quite a lot of things.

(note: functionality to use this was available anyway since the method it invoke was also publicly exposed as static.

### CODE REVIEW

File: src/EFCore/Infrastructure/ExpressionExtensions.cs [hunk 1]
```diff
@@ -6,11 +6,13 @@
...
+using System.Runtime.CompilerServices;
 using JetBrains.Annotations;
 using Microsoft.EntityFrameworkCore.Diagnostics;
 using Microsoft.EntityFrameworkCore.Internal;
 using Microsoft.EntityFrameworkCore.Metadata;
 using Microsoft.EntityFrameworkCore.Query;
+using Microsoft.EntityFrameworkCore.Storage;
 using Microsoft.EntityFrameworkCore.Utilities;
 
 namespace Microsoft.EntityFrameworkCore.Infrastructure
```
Comment (AndriySvyryd) [inline src/EFCore/Infrastructure/ExpressionExtensions.cs]: I wish this class would have been names something like `InternalExpressionExtensions`

Comment (smitpatel) [reply src/EFCore/Infrastructure/ExpressionExtensions.cs]: me too. Such a clash between our own code. 😢

**Comment by smitpatel** [comment]
Is @ajcvickers happy with making this breaking change?

**Comment by ajcvickers** [comment]
It seems okay to me.

/cc @roji @lauxjpn in case it is being used by other providers.

File: src/EFCore/Infrastructure/ExpressionExtensions.cs (hunk 1 above)
Comment (ajcvickers) [reply src/EFCore/Infrastructure/ExpressionExtensions.cs]: Technically we could rename all _internal_ extension method classes without it being a breaking change.

Comment (AndriySvyryd) [reply src/EFCore/Infrastructure/ExpressionExtensions.cs]: but `.Infrastructure` is not `.Internal`

**Pull request merged**